Update Android to use Pusher Beams SDK

mikebridge opened the pull request:
Update to the [pusher beams Client SDK](https://docs.pusher.com/beams/reference/android)

- adds support for `getSubscriptions` and `unsubscribeAll`
- adds ability to process messages while MainActivity is visible (via `setOnMessageReceivedListenerForVisibleActivity`), i.e. you can now use `RNPusherPushNotifications.on("notification", ...)` to subscribe to android notifications while the app is foregrounded.
- adds support for `setOnSubscriptionsChangedListener`

closes #32

NOTES:
- This implements `setOnSubscriptionsChangedListener` as a separate javascript call instead of with a new string event type in `.on()`.  Not sure which is better.
- The status codes on error callbacks are always 0, since it the SDK does not return an HTTP status code.  Additionally, the Pusher Beams SDK [retries failed connections in the background](https://github.com/pusher/push-notifications-android/blob/master/pushnotifications/src/main/java/com/pusher/pushnotifications/api/retrofit2/RequestCallbackWithRetry.kt), so apart from [configuration errors](https://github.com/pusher/push-notifications-android/blob/master/pushnotifications/src/main/java/com/pusher/pushnotifications/PushNotifications.java) it shouldn't return an error.
- setSubscriptions is not implemented, but it could be added with version 1.0.2 which has [this fix](https://github.com/pusher/push-notifications-android/issues/57).

robwalkerco commented:
@mikebridge I'm hoping to use your changes in this PR, but I'm not sure if there are any particular setup steps. It would be great if you could take a look at adding some content to https://github.com/b8ne/react-native-pusher-push-notifications/blob/master/README.md#android :)

mikebridge commented:
@robwalkerco This is from the diff of when we initially pulled this in, but I'm not sure all of it is necessary.  I can clean this up and add it to the PR:

Update `android/app/src/main/AndroidManifest.xml` with:

```xml
<uses-permission android:name="android.permission.WAKE_LOCK" />
<uses-permission android:name="ca.zept.permission.C2D_MESSAGE" />
<uses-permission android:name="android.permission.VIBRATE" />
<uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
```

Add this to `android/app/build.gradle`:

```gradle
dependencies {
   implementation 'com.pusher:push-notifications-android:1.0.1'
}
```

Set up android/app/google-services.json (not sure what all is required for this)

Add `RNPusherPushNotificationsPackage` to `MainApplication.java`:

```java
protected List<ReactPackage> getPackages() {
  return Arrays.<ReactPackage>asList(	  
    new MainReactPackage(),
    new RNPusherPushNotificationsPackage(), // <--- add this
    // ...
}
```

Add to `android/settings.gradle` (below rootProject.name)

```gradle
include ':react-native-pusher-push-notifications'
project(':react-native-pusher-push-notifications').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-pusher-push-notifications/android')
```

Then from typescript:

```typescript
import {Platform} from "react-native";
import {PUSHER_BEAMS_INSTANCE_ID} from "react-native-dotenv";
import RNPusherPushNotifications from "react-native-pusher-push-notifications";

const init= (): void => {
    RNPusherPushNotifications.setInstanceId(PUSHER_BEAMS_INSTANCE_ID);

    RNPusherPushNotifications.on("notification", handleNotification);
    RNPusherPushNotifications.setOnSubscriptionsChangedListener(onSubscriptionsChanged);
};

const subscribe = (interest: string): void => {
    console.log(`Subscribing to "${interest}"`);
    RNPusherPushNotifications.subscribe(
        interest,
        (statusCode, response) => {
            console.error(statusCode, response);
        },
        () => {
            console.log(`CALLBACK: Subscribed to ${interest}`);
        }
    );
};

const handleNotification = (notification: any): void => {
    console.log(notification);
    if (Platform.OS === "ios") {
        console.log("CALLBACK: handleNotification (ios)");
    } else {
        console.log("CALLBACK: handleNotification (android)");
        console.log(notification);
    }
};

const onSubscriptionsChanged = (interests: string[]): void => {
    console.log("CALLBACK: onSubscriptionsChanged");
    console.log(interests);
}
```

robwalkerco commented:
Thanks. I’ll try to get it setup tomorrow using your PR and what you’ve
provided here.

On Thu, 22 Nov 2018 at 17:57, Mike Bridge <notifications@github.com> wrote:

> @robwalkerco <https://github.com/robwalkerco> This is from the diff of
> when we initially pulled this in, but I'm not sure all of it is necessary.
> I can clean this up and add it to the PR:
>
> Update android/app/src/main/AndroidManifest.xml with:
>
> <uses-permission android:name="android.permission.WAKE_LOCK" />
> <uses-permission android:name="ca.zept.permission.C2D_MESSAGE" />
> <uses-permission android:name="android.permission.VIBRATE" />
> <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
>
> Add this to android/app/build.gradle:
>
> dependencies {
>    implementation 'com.pusher:push-notifications-android:1.0.1'
> }
>
> Set up android/app/google-services.json (not sure what all is required for
> this)
>
> Add RNPusherPushNotificationsPackage to MainApplication.java:
>
> protected List<ReactPackage> getPackages() {
>   return Arrays.<ReactPackage>asList(	
>     new MainReactPackage(),
>     new RNPusherPushNotificationsPackage(), // <--- add this
>     // ...
> }
>
> Add to android/settings.gradle (below rootProject.name)
>
> include ':react-native-pusher-push-notifications'
> project(':react-native-pusher-push-notifications').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-pusher-push-notifications/android')
>
> Then from typescript:
>
> import {Platform} from "react-native";import {PUSHER_BEAMS_INSTANCE_ID} from "react-native-dotenv";import RNPusherPushNotifications from "react-native-pusher-push-notifications";
> const init= (): void => {
>     RNPusherPushNotifications.setInstanceId(PUSHER_BEAMS_INSTANCE_ID);
>
>     RNPusherPushNotifications.on("notification", handleNotification);
>     RNPusherPushNotifications.setOnSubscriptionsChangedListener(onSubscriptionsChanged);
> };
> const subscribe = (interest: string): void => {
>     console.log(`Subscribing to "${interest}"`);
>     RNPusherPushNotifications.subscribe(
>         interest,
>         (statusCode, response) => {
>             console.error(statusCode, response);
>         },
>         () => {
>             console.log(`CALLBACK: Subscribed to ${interest}`);
>         }
>     );
> };
> const handleNotification = (notification: any): void => {
>     console.log(notification);
>     if (Platform.OS === "ios") {
>         console.log("CALLBACK: handleNotification (ios)");
>     } else {
>         console.log("CALLBACK: handleNotification (android)");
>         console.log(notification);
>     }
> };
> const onSubscriptionsChanged = (interests: string[]): void => {
>     console.log("CALLBACK: onSubscriptionsChanged");
>     console.log(interests);
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/b8ne/react-native-pusher-push-notifications/pull/33#issuecomment-441095637>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGOPnvQZY1YqrcXfU99Ycm_0iRLi1tB1ks5uxuV6gaJpZM4Yp_of>
> .
>
-- 

Rob Walker
*Co-Founder / CTO*
*Papertrail*
m: +44(0)7732 563087
a: M-Sparc, Menai Science Park, Gaerwen, Anglesey, LL60 6AR
w: www.papertrail.io  e: rob@papertrail.io
<https://www.facebook.com/papertrail.io/>
<https://twitter.com/papertrail_io>
<https://www.linkedin.com/company/kit-log>

robwalkerco commented:
Thanks @mikebridge.
Seems to be working ok so far :)

The only other install step that's required is to add 
`import com.b8ne.RNPusherPushNotifications.RNPusherPushNotificationsPackage;` to MainApplication.java

mikebridge commented:
@robwalkerco Thanks, I've made that edit to the post.

alvincrespo commented:
@mikebridge @robwalkerco Did you run into this error?

```
> Task :app:compileDebugJavaWithJavac FAILED
/Users/alvincrespo/workspace/echobind/truecoach/TruecoachMessenger/android/app/src/main/java/com/truecoachmessenger/MainApplication.java:16: error: package com.b8ne.RNPusherPushNotifications does not exist
import com.b8ne.RNPusherPushNotifications.RNPusherPushNotificationsPackage;
                                         ^
/Users/alvincrespo/workspace/echobind/truecoach/TruecoachMessenger/android/app/src/main/java/com/truecoachmessenger/MainApplication.java:30: error: cannot find symbol
        , new RNPusherPushNotificationsPackage()
              ^
  symbol: class RNPusherPushNotificationsPackage
2 errors


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.
```

mikebridge commented:
@alvincrespo I haven't seen it before but I wonder if I'm missing a step to trigger the installation of the library via gradle, e.g. `yarn build`

mikebridge commented:
@alvincrespo Sorry, that looks like an oversight on my part---I'll add that to the instructions

alvincrespo commented:
@mikebridge Thanks for updating the docs! This is super awesome work. 👍

stephenkiers commented:
Any reason this hasn't been merged yet?

mikebridge commented:
I updated my fork to `push-notifications-android:1.0.2`.  I encountered the error `Could not find firebase-iid-interop.aar`, which I fixed by applying [this](https://github.com/invertase/react-native-firebase/issues/1648).

I'm tagging a new release, 2.2.0.

opula commented:
This works great except I am getting double notifications on android whereas iOS works fine.  Is anyone else experiencing this issue?

mikebridge commented:
@opula I haven't seen that occur on android---do you have some code that reproduces it?

opula commented:
@mikebridge A sample would be a bit hard right now since I ripped out the old pusher code to replace it with this new package.

I can tell you that I called the setInstanceId on the main component constructor but do not call subscribe until after a user logs in.  For your sample, where did you call both methods?

If need be, I'll strip out the extra code for a simple sample.

opula commented:
I've been able to resolve the problem.  Everything now is working as expected.  Along the way, I ran into two issues:

1) Double Notifications - This was because there were some legacy items in my AndroidManifest.xml from the previous pusher notifications implementation.  Once those were removed the notifications worked as expected.

2) Error compiling on debug - I was getting an error that I could not parse google-services.json.  This is because the package name in the google-services.json file must match the app and in debug mode a ".DEBUG" was added.

Just thought I would add in those two issues in case anyone else ran into them.

mikebridge commented:
@opula Great, glad you were able to figure it out!

siers commented:
@mikebridge Hi!

Shouldn't it the main description have this change?

```diff
-    "react-native-pusher-push-notifications": "git+http://git@github.com/ZeptInc/react-native-pusher-push-notifications#2.1.0"
+    "react-native-pusher-push-notifications": "git+http://git@github.com/ZeptInc/react-native-pusher-push-notifications#v2.2.0"
```

gprocell927 commented:
By chance, did you happen to work with the app icon badge count after using this change for Android?   I've been struggling with getting that piece to work without digging into/interfering with this part of it?

mikebridge commented:
@gprocell927 I don't think there is a badge count available on Android.  On ios you can update the badge with the number of notifications, but on Android you just get a little dot.

mikebridge commented:
@siers I had some trouble updating to RN 0.58 on that branch and pushed some versions higher than they probably should be.  I haven't had time to circle back and make a cleaner release.

emilkarl commented:
How is it going? Is there anything I can help with?

emilkarl commented:
Got this working on RN 0.59 with some changes @mikebridge.

1. I my projects build gradle
![image](https://user-images.githubusercontent.com/10711/56664962-6d411a00-66a9-11e9-9baa-3a31c06f23b7.png)

2. in this repo
![image](https://user-images.githubusercontent.com/10711/56665038-995c9b00-66a9-11e9-9156-a48b40a4d0c0.png)

emilkarl commented:
It would be awesome to get this merged! 

Can someone review it and see what needs to be done?

@mikebridge @b8ne @siers @opula @stephenkiers

mikebridge commented:
I'm just testing out a new version with push-notifications-android 1.4.0.  The problems we were having seem to be related to firebase-messaging:17.6.0

emilkarl commented:
Yes you need to use 17.4 I think

mikebridge commented:
17.5 seems to work so far.

emilkarl commented:
Nice
